clippy_lints/src/serde.rs

```diff
@@ -0,0 +1,55 @@
+use rustc::lint::*;
+use rustc::hir::*;
+use utils::{span_lint, get_trait_def_id};
+
+/// **What it does:** This lint checks for mis-uses of the serde API
+///
+/// **Why is this bad?** Serde is very finnicky about how its API should be used, but the type system can't be used to enforce it (yet)
+///
+/// **Known problems:** None.
+///
+/// **Example:** implementing `Visitor::visit_string` but not `Visitor::visit_str`
+declare_lint! {
+    pub SERDE_API_MISUSE, Warn,
+    "Various things that will negatively affect your serde experience"
+}
+
+
+#[derive(Copy, Clone)]
+pub struct Serde;
+
+impl LintPass for Serde {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(SERDE_API_MISUSE)
+    }
+}
+
+impl LateLintPass for Serde {
+    fn check_item(&mut self, cx: &LateContext, item: &Item) {
+        if let ItemImpl(_, _, _, Some(ref trait_ref), _, ref items) = item.node {
+            let did = cx.tcx.expect_def(trait_ref.ref_id).def_id();
+            if let Some(visit_did) = get_trait_def_id(cx, &["serde", "de", "Visitor"]) {
```

Comment:
Does this work if there is a `use serde::de` and they `impl de::Visitor`? Does it work if they `impl ::serde::de::Visitor`? Does it work if they `use serde::de::Visitor as SerdeVisitor`?